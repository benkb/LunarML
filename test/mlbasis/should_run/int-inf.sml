datatype 'a result = VALUE of 'a
                   | OVERFLOW
                   | DOMAIN
fun resultToString (VALUE x) = Int.toString x
  | resultToString OVERFLOW = "Overflow"
  | resultToString DOMAIN = "DOMAIN"
val () = List.app (fn (s, x, expected) =>
                      let val actual = VALUE (IntInf.log2 x)
                                       handle Overflow => OVERFLOW
                                            | Domain => DOMAIN
                      in if expected = actual then
                             print ("log2 " ^ s ^ ": OK\n")
                         else
                             print ("log2 " ^ s ^ ": mismatch (expected " ^ resultToString expected ^ ", but got " ^ resultToString actual ^ ")\n")
                      end
                  )
                  [("0", 0, DOMAIN)
                  ,("~1", ~1, DOMAIN)
                  ,("1", 1, VALUE 0)
                  ,("2", 2, VALUE 1)
                  ,("3", 3, VALUE 1)
                  ,("4", 4, VALUE 2)
                  ,("5", 5, VALUE 2)
                  ,("6", 6, VALUE 2)
                  ,("7", 7, VALUE 2)
                  ,("8", 8, VALUE 3)
                  ,("9", 9, VALUE 3)
                  ,("10", 10, VALUE 3)
                  ,("2^16-1", IntInf.<< (1, 0w16) - 1, VALUE 15)
                  ,("2^16", IntInf.<< (1, 0w16), VALUE 16)
                  ,("2^16+1", IntInf.<< (1, 0w16) + 1, VALUE 16)
                  ,("2^64-1", IntInf.<< (1, 0w64) - 1, VALUE 63)
                  ,("2^64", IntInf.<< (1, 0w64), VALUE 64)
                  ,("2^64+1", IntInf.<< (1, 0w64) + 1, VALUE 64)
                  ,("2^100-1", IntInf.<< (1, 0w100) - 1, VALUE 99)
                  ,("2^100", IntInf.<< (1, 0w100), VALUE 100)
                  ,("2^100+1", IntInf.<< (1, 0w100) + 1, VALUE 100)
                  ,("2^1000-1", IntInf.<< (1, 0w1000) - 1, VALUE 999)
                  ,("2^1000", IntInf.<< (1, 0w1000), VALUE 1000)
                  ,("2^1000+1", IntInf.<< (1, 0w1000) + 1, VALUE 1000)
                  ];
val () = List.app (fn x =>
                      List.app (fn amount =>
                                   let val y = IntInf.~>> (x, amount)
                                   in print (IntInf.toString x ^ " ~>> " ^ Word.fmt StringCvt.DEC amount ^ " = " ^ IntInf.toString y ^ "\n")
                                   end
                               )
                               [0w0, 0w1, 0w2, 0w31, 0w32, 0w63, 0w64, 0w127, 0w128, 0w192, 0wxFFFF_FFFF]
                  )
                  [0, 1, 2, 3, ~1, ~2, ~3, 37, ~42, 999, ~999, IntInf.<< (1, 0w64), IntInf.<< (~1, 0w100) + 3377, IntInf.<< (1, 0w256) + IntInf.<< (1, 0w128)];
