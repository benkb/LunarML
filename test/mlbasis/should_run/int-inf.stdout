log2 0: OK
log2 ~1: OK
log2 1: OK
log2 2: OK
log2 3: OK
log2 4: OK
log2 5: OK
log2 6: OK
log2 7: OK
log2 8: OK
log2 9: OK
log2 10: OK
log2 2^16-1: OK
log2 2^16: OK
log2 2^16+1: OK
log2 2^64-1: OK
log2 2^64: OK
log2 2^64+1: OK
log2 2^100-1: OK
log2 2^100: OK
log2 2^100+1: OK
log2 2^1000-1: OK
log2 2^1000: OK
log2 2^1000+1: OK
0 ~>> 0 = 0
0 ~>> 1 = 0
0 ~>> 2 = 0
0 ~>> 31 = 0
0 ~>> 32 = 0
0 ~>> 63 = 0
0 ~>> 64 = 0
0 ~>> 127 = 0
0 ~>> 128 = 0
0 ~>> 192 = 0
0 ~>> 4294967295 = 0
1 ~>> 0 = 1
1 ~>> 1 = 0
1 ~>> 2 = 0
1 ~>> 31 = 0
1 ~>> 32 = 0
1 ~>> 63 = 0
1 ~>> 64 = 0
1 ~>> 127 = 0
1 ~>> 128 = 0
1 ~>> 192 = 0
1 ~>> 4294967295 = 0
2 ~>> 0 = 2
2 ~>> 1 = 1
2 ~>> 2 = 0
2 ~>> 31 = 0
2 ~>> 32 = 0
2 ~>> 63 = 0
2 ~>> 64 = 0
2 ~>> 127 = 0
2 ~>> 128 = 0
2 ~>> 192 = 0
2 ~>> 4294967295 = 0
3 ~>> 0 = 3
3 ~>> 1 = 1
3 ~>> 2 = 0
3 ~>> 31 = 0
3 ~>> 32 = 0
3 ~>> 63 = 0
3 ~>> 64 = 0
3 ~>> 127 = 0
3 ~>> 128 = 0
3 ~>> 192 = 0
3 ~>> 4294967295 = 0
~1 ~>> 0 = ~1
~1 ~>> 1 = ~1
~1 ~>> 2 = ~1
~1 ~>> 31 = ~1
~1 ~>> 32 = ~1
~1 ~>> 63 = ~1
~1 ~>> 64 = ~1
~1 ~>> 127 = ~1
~1 ~>> 128 = ~1
~1 ~>> 192 = ~1
~1 ~>> 4294967295 = ~1
~2 ~>> 0 = ~2
~2 ~>> 1 = ~1
~2 ~>> 2 = ~1
~2 ~>> 31 = ~1
~2 ~>> 32 = ~1
~2 ~>> 63 = ~1
~2 ~>> 64 = ~1
~2 ~>> 127 = ~1
~2 ~>> 128 = ~1
~2 ~>> 192 = ~1
~2 ~>> 4294967295 = ~1
~3 ~>> 0 = ~3
~3 ~>> 1 = ~2
~3 ~>> 2 = ~1
~3 ~>> 31 = ~1
~3 ~>> 32 = ~1
~3 ~>> 63 = ~1
~3 ~>> 64 = ~1
~3 ~>> 127 = ~1
~3 ~>> 128 = ~1
~3 ~>> 192 = ~1
~3 ~>> 4294967295 = ~1
37 ~>> 0 = 37
37 ~>> 1 = 18
37 ~>> 2 = 9
37 ~>> 31 = 0
37 ~>> 32 = 0
37 ~>> 63 = 0
37 ~>> 64 = 0
37 ~>> 127 = 0
37 ~>> 128 = 0
37 ~>> 192 = 0
37 ~>> 4294967295 = 0
~42 ~>> 0 = ~42
~42 ~>> 1 = ~21
~42 ~>> 2 = ~11
~42 ~>> 31 = ~1
~42 ~>> 32 = ~1
~42 ~>> 63 = ~1
~42 ~>> 64 = ~1
~42 ~>> 127 = ~1
~42 ~>> 128 = ~1
~42 ~>> 192 = ~1
~42 ~>> 4294967295 = ~1
999 ~>> 0 = 999
999 ~>> 1 = 499
999 ~>> 2 = 249
999 ~>> 31 = 0
999 ~>> 32 = 0
999 ~>> 63 = 0
999 ~>> 64 = 0
999 ~>> 127 = 0
999 ~>> 128 = 0
999 ~>> 192 = 0
999 ~>> 4294967295 = 0
~999 ~>> 0 = ~999
~999 ~>> 1 = ~500
~999 ~>> 2 = ~250
~999 ~>> 31 = ~1
~999 ~>> 32 = ~1
~999 ~>> 63 = ~1
~999 ~>> 64 = ~1
~999 ~>> 127 = ~1
~999 ~>> 128 = ~1
~999 ~>> 192 = ~1
~999 ~>> 4294967295 = ~1
18446744073709551616 ~>> 0 = 18446744073709551616
18446744073709551616 ~>> 1 = 9223372036854775808
18446744073709551616 ~>> 2 = 4611686018427387904
18446744073709551616 ~>> 31 = 8589934592
18446744073709551616 ~>> 32 = 4294967296
18446744073709551616 ~>> 63 = 2
18446744073709551616 ~>> 64 = 1
18446744073709551616 ~>> 127 = 0
18446744073709551616 ~>> 128 = 0
18446744073709551616 ~>> 192 = 0
18446744073709551616 ~>> 4294967295 = 0
~1267650600228229401496703201999 ~>> 0 = ~1267650600228229401496703201999
~1267650600228229401496703201999 ~>> 1 = ~633825300114114700748351601000
~1267650600228229401496703201999 ~>> 2 = ~316912650057057350374175800500
~1267650600228229401496703201999 ~>> 31 = ~590295810358705651712
~1267650600228229401496703201999 ~>> 32 = ~295147905179352825856
~1267650600228229401496703201999 ~>> 63 = ~137438953472
~1267650600228229401496703201999 ~>> 64 = ~68719476736
~1267650600228229401496703201999 ~>> 127 = ~1
~1267650600228229401496703201999 ~>> 128 = ~1
~1267650600228229401496703201999 ~>> 192 = ~1
~1267650600228229401496703201999 ~>> 4294967295 = ~1
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 0 = 115792089237316195423570985008687907853610267032561502502920958615344897851392
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 1 = 57896044618658097711785492504343953926805133516280751251460479307672448925696
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 2 = 28948022309329048855892746252171976963402566758140375625730239653836224462848
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 31 = 53919893334301279589334030174039261347432745170109673637394308399104
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 32 = 26959946667150639794667015087019630673716372585054836818697154199552
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 63 = 12554203470773361527671578846415332832241604377075488129024
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 64 = 6277101735386680763835789423207666416120802188537744064512
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 127 = 680564733841876926926749214863536422914
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 128 = 340282366920938463463374607431768211457
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 192 = 18446744073709551616
115792089237316195423570985008687907853610267032561502502920958615344897851392 ~>> 4294967295 = 0
